config/config.go

```diff
@@ -1,675 +1,208 @@
+// Package config uses the exact same precedence order as Viper, where
+// each item takes precedence over the item below it:
+//
+//   - explicit call to Set (case insensitive)
+//   - flag (case insensitive)
+//   - env (case sensitive - see notes below)
+//   - config (case insensitive)
+//   - key/value store (case insensitive)
+//   - default (case insensitive)
+//
+// Environment variable resolution is performed based on the following rules:
+//   - If the key contains only uppercase characters, numbers and underscores, the environment variable is looked up in its entirety, e.g. SOME_VARIABLE -> SOME_VARIABLE
+//   - In all other cases, the environment variable is transformed before being looked up as following:
+//     1. camelCase is converted to snake_case, e.g. someVariable -> some_variable
+//     2. dots (.) are replaced with underscores (_), e.g. some.variable -> some_variable
+//     3. the resulting string is uppercased and prefixed with RSERVER_, e.g. some_variable -> RSERVER_SOME_VARIABLE
 package config
 
 import (
 	"fmt"
-	"os"
 	"regexp"
 	"strconv"
 	"strings"
 	"sync"
 	"time"
 
-	"github.com/fsnotify/fsnotify"
-	"github.com/joho/godotenv"
-	"github.com/spf13/cast"
 	"github.com/spf13/viper"
 )
 
-var matchAllCap = regexp.MustCompile("([a-z0-9])([A-Z])")
+// regular expression matching lowecase letter followed by an uppercase letter
+var camelCaseMatch = regexp.MustCompile("([a-z0-9])([A-Z])")
 
-var (
-	whSchemaVersion        string
-	hotReloadableConfig    map[string][]*ConfigVar
-	nonHotReloadableConfig map[string][]*ConfigVar
-	configVarLock          sync.RWMutex
-)
-
-type ConfigVar struct {
-	value           interface{}
-	multiplier      interface{}
-	isHotReloadable bool
-	defaultValue    interface{}
-	keys            []string
-}
-
-func newConfigVar(value, multiplier, defaultValue interface{}, isHotReloadable bool, keys []string) *ConfigVar {
-	return &ConfigVar{
-		value:           value,
-		multiplier:      multiplier,
-		isHotReloadable: isHotReloadable,
-		defaultValue:    defaultValue,
-		keys:            keys,
-	}
-}
+// regular expression matching uppercase letters contained in environment variable names
+var upperCaseMatch = regexp.MustCompile("^[A-Z0-9_]+$")
 
-func getConfigValue(configVar *ConfigVar) interface{} {
-	return configVar.value
-}
-
-// Rudder server supported config constants
-const (
-	EmbeddedMode      = "embedded"
-	MasterMode        = "master"
-	MasterSlaveMode   = "master_and_slave"
-	SlaveMode         = "slave"
-	OffMode           = "off"
-	PooledWHSlaveMode = "embedded_master"
-)
-
-// TransformKey as package method to get the formatted env from a give string
-func TransformKey(s string) string {
-	snake := matchAllCap.ReplaceAllString(s, "${1}_${2}")
-	snake = strings.ReplaceAll(snake, ".", "_")
-	return "RSERVER_" + strings.ToUpper(snake)
-}
+// default, singleton config instance
+var c *Config
```

Comment:
I borrowed this pattern from [viper](https://github.com/spf13/viper/blob/master/viper.go) itself. This kind of shadowing acts as a safeguard, since you wouldn't ever want to have access to the singleton instance and its methods from within the `Config`'s methods.
But if others believe that this is confusing too, we can rename it to something unique.

Reply:
Personally I don't feel the need for the protection especially if done via shadowing. If possible I'd go for a more descriptive name like `defaultConfig` (or `singletonConfig`, etc...).